decisions/2024-08-30_central_sync_queue_pull_reduction.md

```diff
@@ -0,0 +1,164 @@
+Older versions of sync just used sync queue mechanism. Master lists were filtered down in queuing to only go to sites where that master list was used by a store and master lists that were visible to their customers and suppliers. Or so was the intent at some point - just checking the code it's a bit clapped and appears to send all master lists everywhere in the queue, then for each store re-queues masters lists that are used by active stores and visible stores/names (redundantly, due to everything already being in the queue). In regular activities, any update to any master list goes to all sites.
+
+This could obviously be improved to actually only queue necessary updates to necessary sites.
+
+Remember that OMS uses master lists for item visibility, while Legacy mSupply uses item_store_joins which are made to match their master list based visibility control (backlog task to potentially change this in LCS, probably move to OMS sooner 😉).
+
+One feature that (I'm not sure is implemented in OMS), is to help prevent users from issue stock to customers who shouldn't receive it. For customer mSupply stores, this is items that aren't on their master lists, so are not visible.
+
+## Options
+
+### Option 1 - LCS improve queuing filter and move master lists back to queue rather than change log
+
+This would be contentious, as a key driver for the change log was to prevent spamming the `sync_out` table with updates every time you update a master list, so moving it back to the queue would be coming full circle heh. Some of our bigger users have >1000 sites, so every line touched _could_ require upserting >1000 sync_out records. central_change_log could bring that down to 1 upsert (if all sites were OMR).
+
+But also would work. We could make master list queued data again and improving queuing with filtering to only send updates to where they need to go. This would fix the initialisation problem and also improve legacy mSupply performance.
+
+_Pros:_
+
+- Should improve sync performance over all, including UX on legacy central servers for users doing admin. (There is merit to fixing the master list queuing filtering on LCS regardless of where this KDD goes!)
+- Fixes the problem for existing sites all versions of OMS and legacy apps.
+
+_Cons:_
+
+- Still potentially queuing lots of records to many sites. Some users will have >1000 sites and a general master list used across them with >5000 items. These instances will still suffer millions of queued records that central_change_log set out to fix to begin with!
+- Eventually we want OMS only deployments where OMS controls central data, so a solution in LCS land would require us eventually fixing it again in OMCS land.
+- Doing work in LCS when we're trying to focus on OMS
+
+### Option 2 - LCS central_change_log filtering per log processed
+
+We implement a system that allows sync v5 change log requests to have some filtering on the LCS side.
+
+Suppose a site requests 5000 change_log records from cursor 0:
```

Comment:
What if they aren't querying from cursor 0?

E.g. I've initialised my site, and initially I don't need `Master List A`
Then `Master List A` is added to my site (Update on `master_list_name_join` I think?)

I start my cursor at 1000, and I see the new `master_list_name_join` record but not associated `master_list` would be sent right?

Reply:
ded :( yes it would be borken.

Reply:
This actually affects all the filtering? For instance, it is an occasional activity to move a store from central to an active remote site (or even from another remote site). Currently past records for that store may be earlier in the changelog than the new site's cursor right? I suppose in this case we could query all changelogs by a store_id and bump the sequence number on them to the end of changelog.

Reply:
Yes I think so, if we're filtering out record we'll run into this kind of issue...

Reply:
Such a great point @jmbrunskill.

So if we are filtering by a join (actually similar for name), we have an issue like this. 

I've had a good think of alternative way to do this, and I am struggling to figure out how to do it with logic that I would be ok with, apart from syncing all of that master data everywhere. Even updating master list and lines in changelog to the end when master list name join changes, doesn't sit well with me.

Although this issues is about trying to reduce number of records that are required for initialisation, i fee like it also crosses the question, can we actually sync all master data efficiently if we do some optimisation ?

Reply:
Yea I'm on board with parking and just focusing other optimisations due to this, it's a good catch @jmbrunskill!